handbook/debugging.md

```diff
@@ -0,0 +1,71 @@
+# Debugging
+
+One of the most important programming skills to develop is how to figure out what went wrong with your code.
+
+It can be frustrating to write a bunch of code and immediately get a big error, so here's some advice to bear in mind.
+
+## You're not a bad developer
+
+The most important thing is to remember that getting stuck doesn't say anything about you or your skill as a developer.
+
+It's easy for your first reaction to be "I suck at this, no one else has these problems". _Everyone_ makes typos, forgets how things work or gets inscrutable errors in their terminal. The mark of a more experienced developer isn't avoiding errors entirely, it's reacting to them constructively.
+
+## Understanding the problem
+
+Make sure you stop and understand exactly what went wrong before you start trying to fix it. Otherwise you might go down an entirely wrong path, which will make it harder to figure out what the real original problem was later.
+
+### Read the question
+
+What were you actually trying to do? Do you have a good idea in your head of what the code you wrote _should_ do? If you're working on a specific project feature or workshop problem make sure you've read it properly and understood the edge-cases. If you're following some documentation make sure you've re-read exactly how it works.
+
+E.g. What arguments does this function take? What type of thing does it return? Is it asynchronous?
+
+### Read the error
+
+It can be overwhelming to have 50 lines of error message dumped into the console. It's important to stay calm, take a deep breath and read what it says. Error messages are _usually_ quite useful, as long as you know where to look (hint: start at the top).
+
+Try to figure out _where_ the error is coming from. Sometimes there will be unrelated logs you need to ignore from tools like npm. Find the part that looks like the real error message (this gets easier with practice).
+
+JavaScript errors generally consist of an [error type](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors), an error message, then a stack trace. Let's look at an example error and break it down:
+
+```
+Uncaught TypeError: arr.join is not a function
+    at arrayToString (index.html:11)
+    at run (index.html:14)
+```
+
+The first part tells us that our code encountered a "type error". This means we tried to use a value in an impossible way. E.g. accessing a property on `undefined` instead of an object.
```

Comment:
It may be worth adding parenthetically that this is a *very common* source of error in JS and may be worth eliminating from your research before you do anything else.